tests/llmobs/test_llmobs.py

```diff
@@ -0,0 +1,253 @@
+            pass
+    assert _get_llmobs_parent_id(llm_span) is None
+    with tracer.trace("root_llm_span", span_type=SpanTypes.LLM) as root_span:
+        with tracer.trace("child_span") as child_span:
+            with tracer.trace("llm_span", span_type=SpanTypes.LLM) as grandchild_span:
+                pass
+    assert _get_llmobs_parent_id(root_span) is None
+    assert _get_llmobs_parent_id(child_span) == str(root_span.span_id)
+    assert _get_llmobs_parent_id(grandchild_span) == str(root_span.span_id)
+
+
+class TestSessionId:
+    def test_propagate_from_ancestors(self, tracer):
+        """
+        Test that session_id is propagated from the nearest LLMObs span in the span's ancestor tree
+        if no session_id is not set on the span itself.
+        """
+        with tracer.trace("root_llm_span", span_type=SpanTypes.LLM) as root_span:
+            root_span._set_ctx_item(const.SESSION_ID, "test_session_id")
+            with tracer.trace("child_span"):
+                with tracer.trace("llm_span", span_type=SpanTypes.LLM) as llm_span:
+                    pass
+        assert _get_session_id(llm_span) == "test_session_id"
+
+    def test_if_set_manually(self, tracer):
+        """Test that session_id is extracted from the span if it is already set manually."""
+        with tracer.trace("root_llm_span", span_type=SpanTypes.LLM) as root_span:
+            root_span._set_ctx_item(const.SESSION_ID, "test_session_id")
+            with tracer.trace("child_span"):
+                with tracer.trace("llm_span", span_type=SpanTypes.LLM) as llm_span:
+                    llm_span._set_ctx_item(const.SESSION_ID, "test_different_session_id")
+        assert _get_session_id(llm_span) == "test_different_session_id"
+
+    def test_propagates_ignore_non_llmobs_spans(self, tracer, llmobs_events):
+        """
+        Test that when session_id is not set, we propagate from nearest LLMObs ancestor
+        even if there are non-LLMObs spans in between.
+        """
+        with tracer.trace("root_llm_span", span_type=SpanTypes.LLM) as llm_span:
+            llm_span._set_ctx_item(const.SPAN_KIND, "llm")
+            llm_span._set_ctx_item(const.SESSION_ID, "session-123")
+            with tracer.trace("child_span"):
+                with tracer.trace("llm_grandchild_span", span_type=SpanTypes.LLM) as grandchild_span:
+                    grandchild_span._set_ctx_item(const.SPAN_KIND, "llm")
+                    with tracer.trace("great_grandchild_span", span_type=SpanTypes.LLM) as great_grandchild_span:
+                        great_grandchild_span._set_ctx_item(const.SPAN_KIND, "llm")
+
+        llm_event, grandchild_event, great_grandchild_event = llmobs_events
+        assert llm_event["session_id"] == "session-123"
+        assert grandchild_event["session_id"] == "session-123"
+        assert great_grandchild_event["session_id"] == "session-123"
+
+
+def test_input_value_is_set(tracer, llmobs_events):
```

Comment:
I think there's gonna be some duplication with the below tests and the ones in `test_llmobs_service.py` just because the LLMObs service tests revolve around LLMObs.annotate() which performs the internal logic of these tests here. Up to you if we should remove the tests here and just simplify the ones in `test_llmobs_service.py`, or vice versa

Reply:
definitely down to consolidate the business logic tests from `test_llmobs_service` here as well if that makes sense to you.

I'm thinking the tests in this module (`test_llmobs`) cover the higher level API and semantics and the `test_llmobs_service.py` can keep unit tests for specific components. Then going forward if we refactor things it's clear what tests should be kept.

Reply:
Yup totally fine with that! We'll simplify `test_llmobs_service.py` instead to focus on implementation level logic